python/GafferUI/TextInputDialogue.py

```diff
@@ -40,7 +40,7 @@
 
 class TextInputDialogue( GafferUI.Dialogue ) :
 
-	def __init__( self, initialText="", title="Enter text", cancelLabel="Cancel", confirmLabel="OK", multiLine = False, **kw ) :
+	def __init__( self, initialText="", title="Enter text", cancelLabel="Cancel", confirmLabel="OK", multiLine = False, validator = None, **kw ) :
```

Comment:
The goal with GafferUI is to completely hide the Qt underpinnings from the public API, so expecting a QValidator here isn't really legitimate. I appreciate the intention in terms of improving the user experience though - here are a couple of other possibilities :

1. A [brazen hack](https://github.com/johnhaddon/gaffer/commit/c5a665d3192c130b9372133d03952e99f45be160). The benefit of this approach would be that we're not exposing anything in the public API. If we want to change things in future, it's much easier if we only have to worry about our internal code, and not what other people might be doing with the API.
2. Introduce an official, public, Qt-free validator API.
3. Change `GraphComponent::setName()` so that it never throws, but just replaces invalid characters with `_`. The [documentation](https://github.com/GafferHQ/gaffer/blob/main/include/Gaffer/GraphComponent.h#L102) even suggests that the API is designed to allow that, so I'm not sure why we're not doing it already. This might be worthwhile anyway, but a validator still has the benefit of showing you what you're going to get _before_ you get it.
4. Remove the name constraints from GraphComponent. I originally added them because I wanted to use Python's attribute accessors (`Node.plug`) rather than item accessors (`Node["plug"]`) but that quickly disappeared when I wanted plugs like `in` (that are keywords in Python) and realised that we needed different namespaces for methods and children. I don't know if someone somewhere might be relying on the constraints though. 

I think all those have merit and they're not mutually exclusive, but tend to think that 2-4 are a bit out of scope for this PR. I maybe shouldn't say this, but I think 1) is actually a fairly good short term solution - it achieves what we want for the user, without us needing to contemplate the consequences of any publicly visible changes. I'm sure we'd want to do something cleaner in future, but maybe it's reasonable for now?

Reply:
That sounds reasonable, I went with the solution in 1 along with a todo summarizing 2-4. Fixed, squashed, reorganized a bit and added the `Box` validator in 7c323b7d5d0057f8e52c03a402ac05a2b97f6d8d and 3e2002d08de71366be16e8ce353838d972c309dd